pkg/awsutils/awsutils.go

```diff
@@ -513,75 +524,247 @@ func (cache *EC2InstanceMetadataCache) initWithEC2Metadata(ctx context.Context)
+		return nil
+	}
 
-			attributeInput := &ec2.ModifyNetworkInterfaceAttributeInput{
-				Groups:             sgIDs,
-				NetworkInterfaceId: aws.String(eniID),
-			}
-			start := time.Now()
-			_, err = cache.ec2SVC.ModifyNetworkInterfaceAttribute(context.Background(), attributeInput)
-			prometheusmetrics.Ec2ApiReq.WithLabelValues("ModifyNetworkInterfaceAttribute").Inc()
-			prometheusmetrics.AwsAPILatency.WithLabelValues("ModifyNetworkInterfaceAttribute", fmt.Sprint(err != nil), awsReqStatus(err)).Observe(msSince(start))
+	var eniIDs []string
+	for _, ds := range dsAccess.DataStores {
+		eniIDs = append(eniIDs, cache.getFilteredENIs(ds, true)...) // only secondary subnet ENIs
+	}
 
-			if err != nil {
-				if errors.As(err, &awsAPIError) {
-					if awsAPIError.ErrorCode() == "InvalidNetworkInterfaceID.NotFound" {
-						awsAPIErrInc("IMDSMetaDataOutOfSync", err)
+	cache.applySecurityGroupsToENIs(eniIDs, primarySGs, "Applying primary security groups to")
+
+	return nil
+}
+
+// RefreshSGIDs retrieves security groups
+func (cache *EC2InstanceMetadataCache) RefreshSGIDs(mac string, dsAccess *datastore.DataStoreAccess) error {
+	ctx := context.TODO()
+
+	sgIDs, err := cache.imds.GetSecurityGroupIDs(ctx, mac)
+	if err != nil {
+		awsAPIErrInc("GetSecurityGroupIDs", err)
+		return err
+	}
+
+	addedCount, deletedCount := cache.detectSecurityGroupChanges(sgIDs, &cache.securityGroups, "primary")
+	cache.securityGroups.Set(sgIDs)
+
+	if !cache.useCustomNetworking && (addedCount != 0 || deletedCount != 0) {
+		var eniIDs []string
+
+		// When subnet discovery is enabled, only apply primary SGs to primary subnet ENIs
+		if cache.useSubnetDiscovery {
+			for _, ds := range dsAccess.DataStores {
+				// Get only primary subnet ENIs (onlySecondarySubnets=false)
+				primarySubnetENIs := cache.getFilteredENIs(ds, false)
+				for _, eniID := range primarySubnetENIs {
+					// Filter out unmanaged ENIs
+					if !cache.unmanagedENIs.Has(eniID) {
+						eniIDs = append(eniIDs, eniID)
 					}
 				}
-				checkAPIErrorAndBroadcastEvent(err, "ec2:ModifyNetworkInterfaceAttribute")
-				awsAPIErrInc("ModifyNetworkInterfaceAttribute", err)
-				prometheusmetrics.Ec2ApiErr.WithLabelValues("ModifyNetworkInterfaceAttribute").Inc()
-				// No need to return error here since retry will happen in 30 seconds and also
-				// If update failed due to stale ENI then returning error will prevent updating SG
-				// for following ENIs since the list is sorted
-				log.Debugf("refreshSGIDs: unable to update the ENI %s SG - %v", eniID, err)
 			}
+		} else {
+			// Original behavior: apply to all managed ENIs when subnet discovery is disabled
+			for _, ds := range dsAccess.DataStores {
+				eniInfos := ds.GetENIInfos()
+				for eniID := range eniInfos.ENIs {
+					eniIDs = append(eniIDs, eniID)
+				}
+			}
+
+			newENIs := StringSet{}
+			newENIs.Set(eniIDs)
+			filteredENIs := newENIs.Difference(&cache.unmanagedENIs)
+			eniIDs = filteredENIs.SortedList()
```

Comment:
I saw ENI IDs and SGs being sorted everywhere and not sure why they need to be sorted.